### PR TITLE
Feature ink metadata for substrate

### PR DESCRIPTION
**Pull request by extraymond**
attempt to generate ink compatible metadata using ink_metadata::InkProject
currently relies on unmerged branched of scale-info, ink_metadata.

### CODE REVIEW

**Review by seanyoung — CHANGES_REQUESTED** [review]
Looks great!

Please rebase against latest `main` and add `Signed-off-by: real name <email@address>` to each commit (or even better, squash into one commit)

File: src/abi/substrate.rs
Comment (extraymond) [reply src/abi/substrate.rs:107]: Got it! I couldn't find the place to replace balance types from the ABI, that might need to be configured prior to abi generation.

**Comment by extraymond** [comment]
> Looks great!
> 
> Please rebase against latest `main` and add `Signed-off-by: real name <email@address>` to each commit (or even better, squash into one commit)

Hope it's ok that I force merged --squashed all the changes into a new commit, I couldn't find a way to sign past commits without having to fix all the conflicts.

**Comment by extraymond** [comment]
@seanyoung Adjusted the address_length and variant sort as requested.

**Review by seanyoung — CHANGES_REQUESTED** [review]
There are a bunch of tests which fail with:
```
thread 'substrate_tests::inheritance::inherit_variables' panicked at 'must have at least one message',
```
This is because there is a test contract with only a contructor, no functions. If the ink crate really enforces this, then two things should happen:
- Add dummy function to the tests so the tests pass
- Solang should enforce that on substrate, at least one public function exists and give a nice error message. If someone compiles a contract like so `contract d {}` then we don't want an assertion failure in an ink crate, we want a nice message say "you need at least one function on substrate".

**Comment by xermicus** [comment]
So as discussed in chat we stick with the requirement to have at least one public message.
* Contracts without means to interact with have very niche to no use case
* We do as `ink!` does
* Allows us the reuse the `ink!` crate.

 I'll add a better check and error message in sema.

@extraymond is this OK for you as well? Or do you have a specific use case in mind for contracts without messages?

**Comment by extraymond** [comment]
> So as discussed in chat we stick with the requirement to have at least one public message (substrate only).
> 
>     * Contracts without means to interact with have very niche to no use case
> 
>     * We do as `ink!` does
> 
>     * Allows us the reuse the `ink!` crate.
> 
> 
> I'll add a better check and error message in sema.
> 
> @extraymond is this OK for you as well? Or do you have a specific use case in mind for contracts without messages?

Yeah let's try this first.

**Comment by xermicus** [comment]
I implemented a sema check to catch contracts without public messages. How-ever, it would actually be fine for base contracts. Declaring contracts without public messages as non-concrete some-how could solve the issue elegantly @seanyoung ? The goal  is actually to just prevent metadata generation for those contracts.

File: src/abi/substrate.rs
Comment (seanyoung) [inline src/abi/substrate.rs:110]: I *think* this can be written without a cast: `BigInt::from(ns.address_length).unwrap()`

File: src/sema/mod.rs
Comment (xermicus) [reply src/sema/mod.rs:60]: I see your point and I already tried to implement it more or less like that yesterday. The problem I faced is how would you detect that a contract, while it may not have any public methods, when it is used as a base contract it is fine. So actually we can not decide whether a contract is valid or not by looking at them individually, it would be wrong to `error` or even `warn` about a base contract with no public functions. So instead I came to the conclusion in `sema`, logically it would be correct to warn if there is no public message __at all__.

Or is there an easy and inexpensive way to check whether a contract is used as a base I missed out?

Comment (xermicus) [reply src/sema/mod.rs]: As discussed in our daily, I implemented it now so that it warns the user that any contract without external messages should be declared `abstract`.

This got me thinking, could we actually `error` on that? Only because it is allowed in `solc` does not mean we ultimately must allow it too. If this would result into an error in `sema` we could spare us various redundant checks throughout `codegen` and metadata generation. @seanyoung what do you think?

File: src/sema/contracts.rs
Comment (seanyoung) [inline src/sema/contracts.rs:698]: - lower case contract please
- external or public
-  "messages" is an ink! term, this might be confusing for Solidity developers. I think `public or external functions` makes more sense.
- It might be helpful to explain why this is a problem. Something like "Substrate requires at least one message" or a link to the documentation with a section on this.

File: src/sema/ast.rs
Comment (seanyoung) [inline src/sema/ast.rs:647]: This should check for function type and use all functions, included inherited ones.

```rust
    pub fn has_public_function(&self, ns: &Namespace) -> bool {
        self.all_functions.keys().any(|func_no| {
            let func = &ns.functions[*func_no];

            func.ty == pt::FunctionTy::Function && func.is_public()
        })
    }
```

File: src/sema/contracts.rs
Comment (xermicus) [reply src/sema/contracts.rs:698]: Are you sure this is the right location for this check?

The `data_locations_01` test contains the follow code:
```solidity
contract foo {
    function bar(uint calldata x) public returns () {
  }
}
```

This erroneously fails with for not having public messages, probably because the previous failure (`data location 'calldata' can only be specified for array, struct or mapping`). I'm having a bit of hard time finding the right place for this check to happen...

Comment (seanyoung) [reply src/sema/contracts.rs:698]: So if the prototype of a function can't be resolved it's not added to the ast. This is should really be fixed, which is a large issue to solve (lots of code paths to consider)

However, for now, only perform the check if there are no diagnostic errors so far. So `if !ns.diagnostics.any_error() { ... }`

Comment (xermicus) [reply src/sema/contracts.rs:698]: > So if the prototype of a function can't be resolved it's not added to the ast. This is should really be fixed, which is a large issue to solve (lots of code paths to consider)

I see, thanks!

> However, for now, only perform the check if there are no diagnostic errors so far. So if !ns.diagnostics.any_error() { ... }

Sorry I forgot to mention, I already tried that and it didn't help in this case

File: src/abi/substrate.rs
Comment (seanyoung) [inline src/abi/substrate.rs:59]: Nit: rather than using `format!()` to create a string and then matching the string, you could do:

```rust
ast::Type::Uint(n) => match n.next_power_of_two()) {
        8 => TypeDefPrimitive::U8,
        16 => TypeDefPrimitive::U16,
        32 => TypeDefPrimitive::U32,
        64 => TypeDefPrimitive::U64,
        128 => TypeDefPrimitive::U128,
        256 => TypeDefPrimitive::U256,
},
```

**Review by seanyoung — COMMENTED** [review]
Looks good to me, I have one minor nitpick

File: src/abi/substrate.rs
Comment (extraymond) [reply src/abi/substrate.rs]: Semi adjusted, the sign part is still needed to generate the path name.

Comment (seanyoung) [reply src/abi/substrate.rs]: I don't know why that is needed, you could write it like to so:

```rust
    let def = match ty {
        ast::Type::Uint(bits) => match bits.next_power_of_two() {
            8 => TypeDefPrimitive::U8,
            16 => TypeDefPrimitive::U16,
            32 => TypeDefPrimitive::U32,
            64 => TypeDefPrimitive::U64,
            128 => TypeDefPrimitive::U128,
            256 => TypeDefPrimitive::U256,
            _ => unreachable!(),
        },
        ast::Type::Int(bits) => match bits.next_power_of_two()) {
            8 => TypeDefPrimitive::I8,
            16 => TypeDefPrimitive::I16,
            32 => TypeDefPrimitive::I32,
            64 => TypeDefPrimitive::I64,
            128 => TypeDefPrimitive::I128,
            256 => TypeDefPrimitive::I256,
            _ => unreachable!(),
        },`
        _ => unreachable!(),
    };
```

Comment (seanyoung) [reply src/abi/substrate.rs]: Ah I now see did you need the string later on

```rust
    let path = serde_json::from_value::<Path<PortableForm>>(serde_json::json!(vec![format!(
        "{signed}{scalety}"
    )]))
    .unwrap();
```
Nevermind, maybe the original was nicer. Up to you. Sorry for the bad review comments

Comment (extraymond) [reply src/abi/substrate.rs]: It's okay, I think it's not much different.

**Comment by xermicus** [comment]
@LucasSte @seanyoung you can review this now. I won't merge until I can depend on some `ink!` v4 beta release, but that shouldn't matter.

**Review by xermicus — APPROVED** [review]
LGTM! This fixes like half of our existing substrate integration test suite :eyes: 

Thank you SO much @extraymond for your initiative, initial explorations and continuous feedback! :heart:

File: src/abi/substrate.rs
Comment (seanyoung) [inline src/abi/substrate.rs:401]: maybe this should say something like `// libraries are never in the public interface`

File: src/abi/substrate.rs
Comment (seanyoung) [inline src/abi/substrate.rs:275]: `f_to_constructor` is not great. How about `function_to_constructor` or simply `constructor_spec`

File: src/abi/substrate.rs
Comment (seanyoung) [inline src/abi/substrate.rs:325]: `f_to_message` is not great. How about `function_to_message` or simply `message_spec`

File: src/abi/substrate.rs
Comment (seanyoung) [inline src/abi/substrate.rs:419]: `event_spec` maybe?

File: src/abi/substrate.rs
Comment (seanyoung) [inline src/abi/substrate.rs:507]: I think this function belongs in `tests/substrate.rs`

File: src/abi/substrate.rs
Comment (xermicus) [reply src/abi/substrate.rs]: How could I possibly forget about this. Thanks!

File: src/abi/substrate.rs
Comment (xermicus) [reply src/abi/substrate.rs]: How in the world did I not spot this myself :monkey: :facepalm:

File: tests/substrate.rs
Comment (seanyoung) [inline tests/substrate.rs:1279]: doesn't need to be `pub` (nitpick)

**Review by seanyoung — CHANGES_REQUESTED** [review]

File: src/abi/substrate.rs
Comment (xermicus) [reply src/abi/substrate.rs]: You are right, but actually this is very obvious when looking at the code anyways. Nuked the comment

File: src/sema/namespace.rs
Comment (seanyoung) [reply src/sema/namespace.rs:785]: It's up to you how you want to handle this.

However, you're breaking solidity contracts like this one https://docs.soliditylang.org/en/v0.8.17/types.html#array-members 
```solidity
contract ArrayContract {
    uint[2**20] aLotOfIntegers;
}
```
And you're not explaining why you're breaking this (I have literally no idea). Does this not make for a frustrating developer experience, for us and solidity developers?

Comment (xermicus) [reply src/sema/namespace.rs]: You are :100: right on this one. I think I was testing out something and this just smuggled itself into the PR. It makes no sense and is not needed. Nuked it.

File: src/abi/substrate.rs
Comment (seanyoung) [inline src/abi/substrate.rs:367]: I think `&"return_type".into()` can be replaced with `"return_type"`

File: .gitignore
Comment (seanyoung) [inline .gitignore:6]: Random file change?

File: src/abi/substrate.rs
Comment (seanyoung) [inline src/abi/substrate.rs:145]: substituted

File: src/abi/substrate.rs
Comment (seanyoung) [inline src/abi/substrate.rs:154]: substituted

**Review by seanyoung — CHANGES_REQUESTED** [review]

File: integration/substrate/arrays.sol
Comment (xermicus) [reply integration/substrate/arrays.sol:16]: This particular one is actually, the JS API does not support array dims > 2048

```
 1) Deploy arrays contract and test
       arrays in account storage:
     Error: PortableRegistry: 7: Error extracting {"path":[],"params":[],"def":{"array":{"len":100000024,"type":6}},"docs":[]}: Only support for [Type; <length>], where length <= 2048
```

Besides that, dealing with large arrays like that wouldn't be economic anyways?

File: tests/substrate_tests/arrays.rs
Comment (xermicus) [reply tests/substrate_tests/arrays.rs:108]: Isn't storing such an array only a good idea in solidity where it does not actually use any storage for "empty" array index slots?

File: integration/substrate/arrays.sol
Comment (seanyoung) [reply integration/substrate/arrays.sol:16]: Why would it not be economic? This is a storage array, it's never allocated in memory. Access is per-element (each element of the array uses a storage slot).

Having said that, I don't know what to do about the javascript limitation. You could skip storage elements if they are too large.

File: tests/substrate_tests/arrays.rs
Comment (seanyoung) [reply tests/substrate_tests/arrays.rs:108]: I don't know about ink!, but in Solidity declaring a large storage array without an initializer means no storage is used until it is accessed, i.e. when individual elements are written to. So yes, "empty" array slots do not use any storage (or when you do `delete bigarray[102];`)

Comment (xermicus) [reply tests/substrate_tests/arrays.rs]: changed to `u16::MAX` as a compromise, WDYT?

Comment (xermicus) [reply tests/substrate_tests/arrays.rs]: Reason is max array len for array type defs in the metadata is u32 anyways.

File: integration/substrate/arrays.sol
Comment (xermicus) [reply integration/substrate/arrays.sol]: Sorry I misread, I thought in the test it would actually return the entire `users` array, not just a single one :facepalm: 

Hmm yes but that is a problem I'd rather solve later on, e.g. when a user actually needs this...

File: src/abi/substrate.rs
Comment (LucasSte) [inline src/abi/substrate.rs:95]:
```suggestion
            // substituted to [u8; address_length]
```

File: src/sema/namespace.rs
Comment (LucasSte) [inline src/sema/namespace.rs:763]: Why is this variable necessary?

**Review by LucasSte — CHANGES_REQUESTED** [review]

File: src/sema/namespace.rs
Comment (xermicus) [reply src/sema/namespace.rs:763]: I need this bool inside the closure. But we can't borrow `self` in the closure, because it is borrowed later in the function as well

**Comment by xermicus** [comment]
`ink!` release didn't happen this week. Pinned the dependencies so we can merge as soon as @LucasSte approves

File: integration/substrate/destruct.spec.ts
Comment (LucasSte) [reply integration/substrate/destruct.spec.ts]: They are still here.

File: src/abi/substrate.rs
Comment (LucasSte) [reply src/abi/substrate.rs]: @xermicus you have not changed this.

**Review by LucasSte — CHANGES_REQUESTED** [review]

**Review by LucasSte — APPROVED** [review]

**Pull request merged**
